arwen/host/execution.go

```diff
@@ -163,59 +160,23 @@ func (host *vmHost) ExecuteOnDestContext(input *vmcommon.ContractCallInput) (*vm
 	storage.PushState()
 	storage.SetAddress(runtime.GetSCAddress())
 
-	gasUsed := uint64(0)
-	defer func() {
-		vmOutput = host.finishExecuteOnDestContext(gasUsed, err)
-	}()
-
 	// Perform a value transfer to the called SC. If the execution fails, this
 	// transfer will not persist.
-	err = output.TransferValueOnly(input.RecipientAddr, input.CallerAddr, input.CallValue)
+	err := output.TransferValueOnly(input.RecipientAddr, input.CallerAddr, input.CallValue)
 	if err != nil {
-		return nil, err
+		return host.finishExecuteOnDestContext(0, err)
 	}
 
-	gasUsed, err = host.execute(input)
+	gasUsed, err := host.execute(input)
 	if err != nil {
-		return nil, err
+		return host.finishExecuteOnDestContext(gasUsed, err)
 	}
 
 	err = host.executeCurrentAsyncContext()
```

Comment:
The error is passed directly to `host.finishExecuteOnDestContext()` which will produce the appropriate `VMOutput` according to the error (nil or otherwise).